modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java

```diff
@@ -0,0 +1,180 @@
+            ByteBuffer[] nioBuffers = in.nioBuffers(1024, maxBytesPerGatheringWrite);
+            int nioBufferCnt = in.nioBufferCount();
+
+            if (nioBufferCnt == 0) {// We have something else beside ByteBuffers to write so fallback to normal writes.
+                writeSpinCount -= doWrite0(in);
+            } else {
+                // Zero length buffers are not added to nioBuffers by ChannelOutboundBuffer, so there is no need
+                // to check if the total size of all the buffers is non-zero.
+                ByteBuffer ioBuffer = getIoBuffer();
+                copyBytes(nioBuffers, nioBufferCnt, ioBuffer);
+                ioBuffer.flip();
+
+                int attemptedBytes = ioBuffer.remaining();
+                final int localWrittenBytes = ch.write(ioBuffer);
+                if (localWrittenBytes <= 0) {
+                    incompleteWrite(true);
+                    return;
+                }
+                adjustMaxBytesPerGatheringWrite(attemptedBytes, localWrittenBytes, maxBytesPerGatheringWrite);
+                in.removeBytes(localWrittenBytes);
+                --writeSpinCount;
+            }
+        } while (writeSpinCount > 0);
+
+        incompleteWrite(writeSpinCount < 0);
+    }
+
+    @Override
+    protected int doReadBytes(ByteBuf byteBuf) throws Exception {
+        final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+        allocHandle.attemptedBytesRead(byteBuf.writableBytes());
+        ByteBuffer ioBuffer = getIoBuffer();
+        ByteBuf wrapped = Unpooled.wrappedBuffer(ioBuffer);
+        wrapped.clear();
+        int bytesRead = wrapped.writeBytes(javaChannel(), allocHandle.attemptedBytesRead());
+        if (bytesRead > 0) {
+            byteBuf.writeBytes(wrapped, bytesRead);
+        }
+        return bytesRead;
+    }
+
+    private static ByteBuffer getIoBuffer() {
+        ByteBuffer ioBuffer = CopyBytesSocketChannel.ioBuffer.get();
+        ioBuffer.clear();
+        return ioBuffer;
+    }
+
+    private void adjustMaxBytesPerGatheringWrite(int attempted, int written, int oldMaxBytesPerGatheringWrite) {
+        // By default we track the SO_SNDBUF when ever it is explicitly set. However some OSes may dynamically change
+        // SO_SNDBUF (and other characteristics that determine how much data can be written at once) so we should try
+        // make a best effort to adjust as OS behavior changes.
+        if (attempted == written) {
+            if (attempted << 1 > oldMaxBytesPerGatheringWrite) {
+                writeConfig.setMaxBytesPerGatheringWrite(attempted << 1);
+            }
+        } else if (attempted > MAX_BYTES_PER_GATHERING_WRITE_ATTEMPTED_LOW_THRESHOLD && written < attempted >>> 1) {
+            writeConfig.setMaxBytesPerGatheringWrite(attempted >>> 1);
+        }
+    }
+
+    private static void copyBytes(ByteBuffer[] source, int nioBufferCnt, ByteBuffer destination) {
```

Comment:
Since we know, we're dealing with heap buffers in the `source` only, we could slightly improve efficiency here via something like:

```
    private static void copyBytes(ByteBuffer[] source, int nioBufferCnt, ByteBuffer destination) {
        for (int i = 0; i < nioBufferCnt && destination.hasRemaining(); i++) {
            ByteBuffer buffer = source[i];
            assert buffer.hasArray();
            int nBytesToCopy = Math.min(destination.remaining(), buffer.remaining());
            final int initialPosition = buffer.position();
            destination.put(buffer.array(), buffer.arrayOffset() + initialPosition, nBytesToCopy);
            buffer.position(initialPosition + nBytesToCopy);
        }
    }
```

WDYT?

Reply:
As far as I can tell, this is pretty much what `ByteBuffer#put` is doing minus maybe a limit write. I prefer to not mess with the underlying array offsets if possible.

Reply:
Sure, it was a very minor point :)